DB/src/main/java/io/deephaven/db/v2/sources/SwitchColumnSource.java

```diff
@@ -1,213 +1,275 @@
         super(currentSource.getType(), currentSource.getComponentType());
-        this.currentSource = currentSource;
         this.updateCommitter = new UpdateCommitter<>(this, SwitchColumnSource::clearPrevious);
         this.onPreviousCommitted = onPreviousCommitted;
+        this.currentSource = currentSource;
     }
 
     private void clearPrevious() {
         final ColumnSource<T> captured = prevSource;
-        prevCycle = -1;
+        prevValidityStep = -1;
         prevSource = null;
         if (onPreviousCommitted != null) {
             onPreviousCommitted.accept(captured);
         }
     }
 
+    public void setNewCurrent(ColumnSource<T> newCurrent) {
+        Assert.eq(newCurrent.getType(), "newCurrent.getType()", getType(), "getType()");
+        Assert.eq(newCurrent.getComponentType(), "newCurrent.getComponentType()", getComponentType(), "getComponentType()");
+        prevSource = currentSource;
+        prevValidityStep = LogicalClock.DEFAULT.currentStep();
+        currentSource = newCurrent;
+        updateCommitter.maybeActivate();
+    }
+
     @Override
     public boolean isImmutable() {
         return false;
     }
 
-    private class SwitchFillContext implements FillContext {
-        final FillContext currentContext;
-        final FillContext prevContext;
+    private abstract class SwitchContext<CT extends Context> implements Context {
 
-        private SwitchFillContext(int chunkCapacity, SharedContext sharedContext) {
-            this.currentContext = currentSource.makeFillContext(chunkCapacity, sharedContext);
-            if (prevSource != null) {
-                this.prevContext = prevSource.makeFillContext(chunkCapacity, sharedContext);
-            } else {
-                this.prevContext = null;
-            }
+        final int chunkCapacity;
+        final SharedContext sharedContext;
+
+        private CT currentContext;
+        private CT prevContext;
+
+        private SwitchContext(final int chunkCapacity, final SharedContext sharedContext) {
+            this.chunkCapacity = Require.geqZero(chunkCapacity, "chunkCapacity");
+            this.sharedContext = sharedContext;
+        }
+
+        abstract CT makeContext(@NotNull final ColumnSource innerSource);
+
+        public CT getCurrentContext() {
+            return currentContext == null
+                    ? currentContext = makeContext(currentSource)
+                    : currentContext;
+        }
+
+        public CT getPrevContext() {
+            return prevInvalid()
+                    ? getCurrentContext()
+                    : prevContext == null
+                    ? prevContext = makeContext(prevSource)
+                    : prevContext;
         }
 
         @Override
         public void close() {
-            currentContext.close();
-            if (prevContext != null) {
-                prevContext.close();
+            //noinspection EmptyTryBlock
```

Comment:
This seems like a very round about way of closing things and null checking.

Reply:
As discussed, it's just better in every way except clarity to the uninitiated.